File: src/features/accounts/schemas/account.schema.ts
import { z } from "zod"

import type { Translate } from "../../../i18n/context"
import { compareDecimals } from "../../../lib/financial-calculations/decimal"
import {
  accountTypeCodes,
  bankSubtypeCodes,
  businessTypeCodes,
  currencyCodes,
  industryCodes,
  investmentTypeCodes,
  metalTypeCodes,
  propertyTypeCodes,
  purityCodes,
  type AccountFormValues,
} from "../types/account-form"

const decimalAmountPattern = /^\d{1,18}(?:\.\d{1,2})?$/
const percentagePattern = /^\d{1,3}(?:\.\d{1,2})?$/
const isoDatePattern = /^\d{4}-\d{2}-\d{2}$/

function validateValuationDate(
  value: string,
  ctx: z.RefinementCtx,
  t: Translate,
) {
  if (!isoDatePattern.test(value)) {
    ctx.addIssue({
      code: "custom",
      path: ["valuationDate"],
      message: t("accounts.validation.valuationDateRequired"),
    })
    return
  }
  if (value > new Date().toISOString().slice(0, 10)) {
    ctx.addIssue({
      code: "custom",
      path: ["valuationDate"],
      message: t("accounts.validation.valuationDateFuture"),
    })
  }
}

export function createAccountSchema(
  t: Translate,
  mode: "create" | "edit" = "create"
): z.ZodType<AccountFormValues, AccountFormValues> {
  return z
    .object({
      name: z.string().trim(),
      accountTypeCode: z.enum(accountTypeCodes),
      currencyCode: z.enum(currencyCodes),
      openingBalance: z.string().trim(),
      bankSubtype: z.union([z.enum(bankSubtypeCodes), z.literal("")]),
      creditCardLimit: z.string().trim(),
      dueDayOfMonth: z.string().trim(),
      investmentType: z.union([z.enum(investmentTypeCodes), z.literal("")]),
      balanceGrams: z.string().trim(),
      propertyType: z.union([z.enum(propertyTypeCodes), z.literal("")]),
      ownershipPercentage: z.string().trim(),
      businessType: z.union([z.enum(businessTypeCodes), z.literal("")]),
      businessTypeOther: z.string().trim(),
      industry: z.union([z.enum(industryCodes), z.literal("")]),
      industryOther: z.string().trim(),
      location: z.string().trim(),
      valuationDate: z.string().trim(),
      valuationMethod: z.string().trim(),
      valuationNotes: z.string().trim(),
      metalType: z.union([z.enum(metalTypeCodes), z.literal("")]),
      purity: z.union([z.enum(purityCodes), z.literal("")]),
      purchaseDate: z.string().trim(),
      costPerUnit: z.string().trim(),
      notes: z.string().trim(),
      isActive: z.boolean(),
    })
    .superRefine((values, ctx) => {
      const requireBalance = (message: string) => {
        if (!values.openingBalance) {
          ctx.addIssue({
            code: "custom",
            path: ["openingBalance"],
            message: t("accounts.validation.balanceRequired"),
          })
        } else if (!decimalAmountPattern.test(values.openingBalance)) {
          ctx.addIssue({ code: "custom", path: ["openingBalance"], message })
        }
      }
      const requirePercentage = () => {
        if (!values.ownershipPercentage) {
          ctx.addIssue({
            code: "custom",
            path: ["ownershipPercentage"],
            message: t("accounts.validation.ownershipPercentageRequired"),
          })
          return
        }
        if (
          !percentagePattern.test(values.ownershipPercentage) ||
          Number(values.ownershipPercentage) > 100
        ) {
          ctx.addIssue({
            code: "custom",
            path: ["ownershipPercentage"],
            message: t("accounts.validation.ownershipPercentageInvalid"),
          })
        }
      }

      if (values.accountTypeCode !== "gold" && !values.name) {
        ctx.addIssue({
          code: "custom",
          path: ["name"],
          message: t("accounts.validation.nameRequired"),
        })
      }

      switch (values.accountTypeCode) {
        case "cash":
        case "other": {
          requireBalance(t("accounts.validation.balanceInvalid"))
          break
        }
        case "bank": {
          requireBalance(t("accounts.validation.balanceInvalid"))
          if (!values.bankSubtype) {
            ctx.addIssue({
              code: "custom",
              path: ["bankSubtype"],
              message: t("accounts.validation.bankSubtypeRequired"),
            })
          }
          if (values.bankSubtype === "credit") {
            if (
              !decimalAmountPattern.test(values.creditCardLimit) ||
              compareDecimals(values.creditCardLimit, "0") !== 1
            ) {
              ctx.addIssue({
                code: "custom",
                path: ["creditCardLimit"],
                message: t("accounts.validation.creditCardLimitInvalid"),
              })
            }
            if (
              values.dueDayOfMonth &&
              !/^(?:[1-9]|[12]\d|3[01])$/.test(values.dueDayOfMonth)
            ) {
              ctx.addIssue({
                code: "custom",
                path: ["dueDayOfMonth"],
                message: t("accounts.validation.dueDayOfMonthInvalid"),
              })
            }
            if (
              decimalAmountPattern.test(values.openingBalance) &&
              decimalAmountPattern.test(values.creditCardLimit) &&
              compareDecimals(values.openingBalance, values.creditCardLimit) ===
                1
            ) {
              ctx.addIssue({
                code: "custom",
                path: ["openingBalance"],
                message: t("accounts.validation.creditBalanceExceedsLimit"),
              })
            }
          }
          break
        }
        case "brokerage": {
          requireBalance(t("accounts.validation.balanceInvalid"))
          if (!values.investmentType) {
            ctx.addIssue({
              code: "custom",
              path: ["investmentType"],
              message: t("accounts.validation.investmentTypeRequired"),
            })
          }
          break
        }
        case "gold": {
          if (!values.metalType) {
            ctx.addIssue({
              code: "custom",
              path: ["metalType"],
              message: t("accounts.validation.metalTypeRequired"),
            })
          }
          break
        }
        case "real_estate": {
          if (mode === "create") requireBalance(t("accounts.validation.balanceInvalid"))
          requirePercentage()
          if (!values.propertyType) {
            ctx.addIssue({
              code: "custom",
              path: ["propertyType"],
              message: t("accounts.validation.propertyTypeRequired"),
            })
          }
          if (mode === "create") validateValuationDate(values.valuationDate, ctx, t)
          break
        }
        case "business": {
          if (mode === "create") requireBalance(t("accounts.validation.balanceInvalid"))
          requirePercentage()
          if (!values.businessType) {
            ctx.addIssue({
              code: "custom",
              path: ["businessType"],
              message: t("accounts.validation.businessTypeRequired"),
            })
          }
          if (values.businessType === "other" && !values.businessTypeOther) {
            ctx.addIssue({
              code: "custom",
              path: ["businessTypeOther"],
              message: t("accounts.validation.businessTypeOtherRequired"),
            })
          }
          if (!values.industry) {
            ctx.addIssue({
              code: "custom",
              path: ["industry"],
              message: t("accounts.validation.industryRequired"),
            })
          }
          if (values.industry === "other" && !values.industryOther) {
            ctx.addIssue({
              code: "custom",
              path: ["industryOther"],
              message: t("accounts.validation.industryOtherRequired"),
            })
          }
          if (mode === "create") validateValuationDate(values.valuationDate, ctx, t)
          break
        }
      }
    }) as z.ZodType<AccountFormValues, AccountFormValues>
}

File: apps/mobile/lib/accounts/account_schema.dart
// Account form validation — 1:1 port of the web Zod `createAccountSchema`
// superRefine (src/features/accounts/schemas/account.schema.ts, docs/accounts.md §3).
// Returns a field -> message map; empty means valid. Errors surface only after
// the first submit attempt (the form gates on that, not this function).

import '../core/decimals.dart';
import 'account_form.dart';
import 'account_models.dart';

final _decimalAmount = RegExp(r'^\d{1,18}(?:\.\d{1,2})?$');
final _percentage = RegExp(r'^\d{1,3}(?:\.\d{1,2})?$');
final _isoDate = RegExp(r'^\d{4}-\d{2}-\d{2}$');
final _dueDay = RegExp(r'^(?:[1-9]|[12]\d|3[01])$');

// Messages are 1:1 with the web i18n `accounts.validation.*` strings.
const _msg = {
  'nameRequired': 'Name is required',
  'balanceRequired': 'Balance is required',
  'balanceInvalid': 'Enter a non-negative amount with up to 2 decimal places',
  'bankSubtypeRequired': 'Select debit or credit',
  'creditCardLimitInvalid':
      'Enter a positive credit card limit with up to 2 decimal places',
  'dueDayOfMonthInvalid': 'Select a due day from 1 through 31',
  'creditBalanceExceedsLimit':
      'Current balance cannot exceed the credit card limit',
  'investmentTypeRequired': 'Select a type of investment',
  'metalTypeRequired': 'Select gold or silver',
  'ownershipRequired': 'Ownership percentage is required',
  'ownershipInvalid': 'Enter a value between 0 and 100',
  'propertyTypeRequired': 'Select a property type',
  'businessTypeRequired': 'Business type is required',
  'businessTypeOtherRequired': 'Specify the business type',
  'industryRequired': 'Industry is required',
  'industryOtherRequired': 'Specify the industry',
  'valuationDateRequired': 'Valuation date is required',
  'valuationDateFuture': 'Valuation date cannot be in the future',
};

Map<String, String> validateAccountForm(
  AccountFormValues v, {
  required bool isCreate,
}) {
  final errors = <String, String>{};

  void requireBalance() {
    if (v.openingBalance.trim().isEmpty) {
      errors['openingBalance'] = _msg['balanceRequired']!;
    } else if (!_decimalAmount.hasMatch(v.openingBalance.trim())) {
      errors['openingBalance'] = _msg['balanceInvalid']!;
    }
  }

  void requirePercentage() {
    final p = v.ownershipPercentage.trim();
    if (p.isEmpty) {
      errors['ownershipPercentage'] = _msg['ownershipRequired']!;
      return;
    }
    if (!_percentage.hasMatch(p) || (double.tryParse(p) ?? 101) > 100) {
      errors['ownershipPercentage'] = _msg['ownershipInvalid']!;
    }
  }

  void validateValuationDate() {
    final d = v.valuationDate.trim();
    if (!_isoDate.hasMatch(d)) {
      errors['valuationDate'] = _msg['valuationDateRequired']!;
      return;
    }
    if (d.compareTo(DateTime.now().toIso8601String().substring(0, 10)) > 0) {
      errors['valuationDate'] = _msg['valuationDateFuture']!;
    }
  }

  if (v.type != AccountType.gold && v.name.trim().isEmpty) {
    errors['name'] = _msg['nameRequired']!;
  }

  switch (v.type) {
    case AccountType.cash:
    case AccountType.other:
      requireBalance();
    case AccountType.bank:
      requireBalance();
      if (v.bankSubtype.isEmpty) {
        errors['bankSubtype'] = _msg['bankSubtypeRequired']!;
      }
      if (v.bankSubtype == 'credit') {
        final limit = v.creditCardLimit.trim();
        if (!_decimalAmount.hasMatch(limit) ||
            (D.compare(limit, '0') ?? -1) <= 0) {
          errors['creditCardLimit'] = _msg['creditCardLimitInvalid']!;
        }
        if (v.dueDayOfMonth.isNotEmpty &&
            !_dueDay.hasMatch(v.dueDayOfMonth.trim())) {
          errors['dueDayOfMonth'] = _msg['dueDayOfMonthInvalid']!;
        }
        if (_decimalAmount.hasMatch(v.openingBalance.trim()) &&
            _decimalAmount.hasMatch(limit) &&
            (D.compare(v.openingBalance.trim(), limit) ?? 0) > 0) {
          errors['openingBalance'] = _msg['creditBalanceExceedsLimit']!;
        }
      }
    case AccountType.brokerage:
      requireBalance();
      if (v.investmentType.isEmpty) {
        errors['investmentType'] = _msg['investmentTypeRequired']!;
      }
    case AccountType.gold:
      if (v.metalType.isEmpty) {
        errors['metalType'] = _msg['metalTypeRequired']!;
      }
    case AccountType.realEstate:
      if (isCreate) requireBalance();
      requirePercentage();
      if (v.propertyType.isEmpty) {
        errors['propertyType'] = _msg['propertyTypeRequired']!;
      }
      if (isCreate) validateValuationDate();
    case AccountType.business:
      if (isCreate) requireBalance();
      requirePercentage();
      if (v.businessType.isEmpty) {
        errors['businessType'] = _msg['businessTypeRequired']!;
      }
      if (v.businessType == 'other' && v.businessTypeOther.trim().isEmpty) {
        errors['businessTypeOther'] = _msg['businessTypeOtherRequired']!;
      }
      if (v.industry.isEmpty) {
        errors['industry'] = _msg['industryRequired']!;
      }
      if (v.industry == 'other' && v.industryOther.trim().isEmpty) {
        errors['industryOther'] = _msg['industryOtherRequired']!;
      }
      // `valuationMethod` is a free-text field (web parity) — no validation.
      if (isCreate) validateValuationDate();
  }

  return errors;
}
